[JENKINS-20187] - Handle growing files when creating a tar file.

alvarolobato opened the pull request:
[JENKINS-20187](https://issues.jenkins-ci.org/browse/JENKINS-20187)

If the file grows while being included on the tar file (typical case of a log file), it will only be included the original size when the operation was started and included on the header of the tar entry.

Removed dependency from `org.apache.tools.tar.TarConstants` in favor of `org.apache.commons.compress.archivers.tar.TarConstants`

@reviewbybees esp @jglick @jtnord @amuniz

alvarolobato committed:
[JENKINS-20187] - Handle growing files when creating a tar file.

reviewbybees commented:
This pull request originates from a [CloudBees](https://www.cloudbees.com/) employee. At CloudBees, we require that all pull requests be reviewed by other CloudBees employees before we seek to have the change accepted. If you want to learn more about our process please see [this explanation](https://github.com/reviewbybees/about#about-reviewbybees).

jglick commented on core/src/main/java/hudson/util/io/TarArchiver.java:
```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
+                    in.close();
+                }
             }
+        } finally { // always close the entry
```
I do not think this needs to be in a `finally`-block. Either there was an error, in which case the tarball is trash, or there was not, in which case it is closed in the normal course of things. The tarball output stream is closed elsewhere I suppose.

alvarolobato replied:
Yes, the tar get's closed else were, but in some use cases you will get an `java.io.IOException: This archives contains unclosed entries` when closing the tar in a finally that will mask the original exception. This way we try not to cause another exception. ie: [FilePath:2288](https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/FilePath.java#L2288-L2291)

jglick commented:
:bee:

oleg-nenashev commented:
Honestly I don't see a reason to perform such graceful handling. In many cases it will also lead to the truncated & invalid files.

I'd vote for just failing file archiving if it changes during the operation.

alvarolobato commented on core/src/main/java/hudson/util/io/TarArchiver.java:
```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
+                        }
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
```
The reason to log is the same as the comment below, there are some cases were you get another exception an loose the actual exception.

jglick replied:
It would be better to use `Throwable.addSuppressed` then.

jtnord replied:
I believe the case here was from the fact that if this errors, then the `in.close` 3 lines below in the finally block will almost certainly throw an exception due to an un-closed stream which will then just walk all over this exception

Changing this to a try-with-resources should fix that issue, as well as making the code cleaner, then just go with Jesse's suggestion of catch, wrap and re-throw.

jglick replied:
Sure, if the code idiom _can_ be converted to `try`-with-resources, then by all means do it.

BTW some Java IDEs offer this conversion as a hint.

alvarolobato replied:
Actually the exception is thrown on the final of the caller not on this one
(the reference I put before) and only when we don't have the final to close
the entry. I'll convert it to try with resources, and see if we can use
Jesse's suggestion.
El 22/6/2016 20:13, "Jesse Glick" <notifications@github.com> escribió:

> In core/src/main/java/hudson/util/io/TarArchiver.java
> <https://github.com/jenkinsci/jenkins/pull/2419#discussion_r68105507>:
>
> > +                FileInputStream in = new FileInputStream(file);
> > +                try {
> > +                    int len;
> > +                    int read = 0;
> > +                    while ((len = in.read(buf)) >= 0) {
> > +                        if (len + read <= size) { // ensure we don't write more bytes than the declared when we created
> > +                                                  // the entry
> > +                            tar.write(buf, 0, len);
> > +                            read += len;
> > +                        } else {
> > +                            tar.write(buf, 0, (int) size - read);
> > +                            break;
> > +                        }
> > +                    }
> > +                } catch (IOException e) {// log the exception in any case
> > +                    LOGGER.log(Level.SEVERE, "Error writing to tar file from: " + file, e);
>
> Sure, if the code idiom *can* be converted to try-with-resources, then by
> all means do it.
>
> BTW some Java IDEs offer this conversion as a hint.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jenkinsci/jenkins/pull/2419/files/64b039bc7a79590877c1800a9aa1bb23549a214f#r68105507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AQXNBPHtA-FuIO2fKadsRtDHxnCRfZzIks5qOXtAgaJpZM4I72ij>
> .
>

alvarolobato commented:
@oleg-nenashev the reason is not graceful handling, is getting the actual exception to the caller.

jglick commented on core/src/main/java/hudson/util/io/TarArchiver.java:
```diff
@@ -87,27 +90,46 @@ public void visit(File file, String relativePath) throws IOException {
+                            read += len;
+                        } else {
+                            tar.write(buf, 0, (int) size - read);
+                            break;
```
Easier to use `BoundedInputStream`.

alvarolobato committed:
- Introduce try-with-resources
- Remove exception login
- Throw new exception with filename and addSuppresed

alvarolobato commented:
I've changed the try for try-with resources and also added the AddSuppresed. As I suspected if I don't close the entry, even if we add the suppressed, the exception will be masked by the tar.close on the caller, we will get this without closing the entry:
``java.io.IOException: This archives contains unclosed entries.
	at hudson.util.io.TarArchiverTest.growingFileTar(TarArchiverTest.java:132)``

instead of this (if closing the entry):
``hudson.os.PosixException: native error calling stat: No such file or directory non_existent_file
	at hudson.util.io.TarArchiverTest.growingFileTar(TarArchiverTest.java:132)``

oleg-nenashev commented:
Anyway it's better than the original behavior. Hence 🐝

alvarolobato commented:
@reviewbybees done

oleg-nenashev commented:
I'm going to merge it today if there is no -1s

jglick commented on core/src/main/java/hudson/util/io/TarArchiver.java:
```diff
@@ -87,23 +91,30 @@ public void visit(File file, String relativePath) throws IOException {
+        try {
+            if (!file.isDirectory()) {
+                // ensure we don't write more bytes than the declared when we created the entry
+                try (BoundedInputStream in = new BoundedInputStream(new FileInputStream(file), size)) {
```
:ant: safer is to use

```java
try (InputStream is = new FileInputStream(file); BoundedInputStream in = new BoundedInputStream(is, size)) {
```

so that you can be guaranteed the `FileInputStream` is closed.

jglick commented on core/src/main/java/hudson/util/io/TarArchiver.java:
```diff
@@ -87,23 +91,30 @@ public void visit(File file, String relativePath) throws IOException {
+                    }
+                } catch (IOException e) {// log the exception in any case
+                    IOException ioE = new IOException("Error writing to tar file from: " + file, e);
+                    ioE.addSuppressed(e);
```
:ant: Huh? Why would you call `addSuppressed` on something which is already the `cause`? This makes no sense to me.

jglick commented:
:bee:

reviewbybees commented:
This pull request has completed our internal processes and we now respectfully request the maintainers of this repository to consider our proposal contained within this pull request for merging.

oleg-nenashev commented:
@reviewbybees done
@alvarolobato Do you want to address @jglick's comments?

alvarolobato commented:
Yes, let me a few hours.

alvarolobato committed:
Address jglick comments

alvarolobato commented:
@oleg-nenashev, I've addressed @jglick  comments.

oleg-nenashev commented:
re-:bee: ready to merge
cc @jenkinsci/code-reviewers
